File: mcp-servers/site-crawler/src/auth/credential-resolver.ts
/**
 * Profile-based credential resolution (TypeScript twin of
 * mcp-servers/aep-core/src/aep_core/auth/credential_resolver.py — see that
 * file for the full policy writeup; this keeps the same contract).
 *
 * MCP tools in this server accept a `profile` name string and a `domain`,
 * never raw secret values. `resolveProfile()` is the only function allowed
 * to turn a profile name into usable credentials, and the result must
 * never be returned from a tool's response payload, logged, or otherwise
 * exposed back to the calling model — it is for internal use (e.g.
 * authenticating a Playwright request) only.
 */

import { readFileSync, existsSync } from "node:fs";
import { resolve } from "node:path";

export class ProfileNotFoundError extends Error {}
export class DomainNotAllowedError extends Error {}
export class SecretBackendError extends Error {}

export interface ResolvedCredentials {
  profileName: string;
  /** Redacted on purpose — never populate with a real secret in logs/toString. */
  toString(): string;
  [key: string]: unknown;
}

interface DopplerConfigRef {
  project: string;
  config: string;
}

interface ProfileMetadata {
  vaultRef?: string;
  dopplerConfig?: DopplerConfigRef;
  allowedDomains: string[];
}

type ProfilesManifest = Record<string, ProfileMetadata>;

const DOPPLER_API_BASE = "https://api.doppler.com/v3";

function profilesManifestPath(): string {
  const override = process.env.MCP_PROFILES_PATH;
  if (override) return override;
  // repo-root/secrets/profiles.json relative to this file at runtime
  return resolve(__dirname, "..", "..", "..", "..", "secrets", "profiles.json");
}

function loadProfilesManifest(): ProfilesManifest {
  const path = profilesManifestPath();
  if (!existsSync(path)) {
    throw new SecretBackendError(
      `Profiles manifest not found at ${path}. Copy secrets/profiles.example.json ` +
        `to secrets/profiles.json and fill it in for local development.`
    );
  }
  return JSON.parse(readFileSync(path, "utf-8")) as ProfilesManifest;
}

function envVarFor(profileName: string, field: string): string {
  const normalized = profileName.toUpperCase().replace(/-/g, "_");
  return `SITE_CRAWLER_PROFILE_${normalized}_${field.toUpperCase()}`;
}

function redactedCreds(profileName: string, fields: Record<string, string>): ResolvedCredentials {
  return {
    profileName,
    ...fields,
    toString: () => `ResolvedCredentials(profileName=${profileName}, <redacted>)`,
  };
}

function resolveLocal(profileName: string): ResolvedCredentials {
  const requiredFields = ["basicAuthUser", "basicAuthPassword"];
  const values: Record<string, string> = {};
  const missing: string[] = [];

  for (const field of requiredFields) {
    const envName = envVarFor(profileName, field);
    const value = process.env[envName];
    if (!value) {
      missing.push(envName);
      continue;
    }
    values[field] = value;
  }

  if (missing.length > 0) {
    throw new SecretBackendError(
      `Missing local secret env vars for profile '${profileName}': ${missing.join(", ")}. ` +
        `Set them, or switch MCP_SECRETS_BACKEND=doppler (see docs/security-model.md).`
    );
  }

  return redactedCreds(profileName, values);
}

function resolveVault(profileName: string, meta: ProfileMetadata): ResolvedCredentials {
  if (!meta.vaultRef) {
    throw new SecretBackendError(`Profile '${profileName}' has no vaultRef configured.`);
  }
  throw new Error(
    `NotImplemented: wire up Vault/AWS Secrets Manager lookup for vaultRef=${meta.vaultRef} here`
  );
}

/**
 * Fetch the full secrets map for one Doppler project/config. Isolated as
 * its own function so tests can stub `globalThis.fetch` instead of
 * hitting the real Doppler API.
 */
async function fetchDopplerSecrets(project: string, config: string, token: string): Promise<Record<string, string>> {
  const url = new URL(`${DOPPLER_API_BASE}/configs/config/secrets/download`);
  url.searchParams.set("project", project);
  url.searchParams.set("config", config);
  url.searchParams.set("format", "json");

  let response: Response;
  try {
    response = await fetch(url, {
      // Doppler API: bearer token as basic-auth username, empty password.
      headers: { Authorization: `Basic ${Buffer.from(`${token}:`).toString("base64")}` },
    });
  } catch (err) {
    throw new SecretBackendError(
      `Doppler request failed for project=${project} config=${config}: ${(err as Error).message}`
    );
  }

  if (!response.ok) {
    throw new SecretBackendError(
      `Doppler request failed for project=${project} config=${config}: HTTP ${response.status}`
    );
  }
  return (await response.json()) as Record<string, string>;
}

async function resolveDoppler(profileName: string, meta: ProfileMetadata): Promise<ResolvedCredentials> {
  const dopplerConfig = meta.dopplerConfig;
  if (!dopplerConfig?.project || !dopplerConfig?.config) {
    throw new SecretBackendError(`Profile '${profileName}' has no dopplerConfig (project/config) configured.`);
  }

  const token = process.env.DOPPLER_TOKEN;
  if (!token) {
    throw new SecretBackendError(
      "DOPPLER_TOKEN is not set. This is the one bootstrap credential this process needs — " +
        "a Doppler service token scoped to the project/config referenced by each profile's " +
        "dopplerConfig. See docs/security-model.md."
    );
  }

  const secretsMap = await fetchDopplerSecrets(dopplerConfig.project, dopplerConfig.config, token);

  const requiredFields = ["basicAuthUser", "basicAuthPassword"];
  const values: Record<string, string> = {};
  const missing: string[] = [];

  for (const field of requiredFields) {
    const secretName = envVarFor(profileName, field);
    const value = secretsMap[secretName];
    if (!value) {
      missing.push(secretName);
      continue;
    }
    values[field] = value;
  }

  if (missing.length > 0) {
    throw new SecretBackendError(
      `Missing Doppler secrets for profile '${profileName}' in ` +
        `${dopplerConfig.project}/${dopplerConfig.config}: ${missing.join(", ")}.`
    );
  }

  return redactedCreds(profileName, values);
}

/**
 * Resolve a profile name + target domain into usable credentials for the
 * site crawler (e.g. HTTP basic auth for a staging environment gated
 * behind auth). Validates `domain` against the profile's `allowedDomains`
 * so a credential can't be replayed against an unintended target.
 */
export async function resolveProfile(profileName: string, domain: string): Promise<ResolvedCredentials> {
  const manifest = loadProfilesManifest();
  const meta = manifest[profileName];
  if (!meta) {
    throw new ProfileNotFoundError(`Unknown profile: ${profileName}`);
  }

  const allowedDomains = meta.allowedDomains ?? [];
  if (!allowedDomains.includes(domain) && !allowedDomains.includes("*")) {
    throw new DomainNotAllowedError(
      `Profile ${profileName} is not authorized for domain ${domain}. Allowed: ${allowedDomains.join(", ")}`
    );
  }

  const backend = (process.env.MCP_SECRETS_BACKEND ?? "local").toLowerCase();
  if (backend === "local") return resolveLocal(profileName);
  if (backend === "vault") return resolveVault(profileName, meta);
  if (backend === "doppler") return resolveDoppler(profileName, meta);

  throw new SecretBackendError(`Unknown MCP_SECRETS_BACKEND: ${backend}`);
}

File: mcp-servers/aep-core/src/aep_core/auth/credential_resolver.py
"""Profile-based credential resolution.

Every MCP tool in this server that touches a real AEP tenant accepts a
`profile` *name* (e.g. "cit_bank_test_user"), never a raw secret. This
module is the only place allowed to turn that name into an actual
credential set, and it does so at call time, inside this process — the
resolved credentials are handed to internal HTTP client code and are never
returned as part of an MCP tool's response payload, logged, or echoed back
to the calling model.

Backend selection is controlled by the `MCP_SECRETS_BACKEND` env var:

  - "local" (default): read `secrets/profiles.json` (gitignored) for the
    profile's metadata (vault ref placeholder + allowedDomains), then pull
    the actual secret values from per-profile environment variables. This
    is the local-dev path only.
  - "vault": production path. Placeholder — wire this up to HashiCorp
    Vault / AWS Secrets Manager. The profile's `vaultRef` from
    profiles.json tells us where to look; nothing here should ever read a
    plaintext secret from a checked-in file in that mode.

See docs/security-model.md for the full policy this module implements.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass
from pathlib import Path
from typing import Any


class ProfileNotFoundError(Exception):
    """Raised when a requested profile does not exist in the manifest."""


class DomainNotAllowedError(Exception):
    """Raised when a profile is used against a domain outside its allowlist."""


class SecretBackendError(Exception):
    """Raised when the configured secrets backend cannot resolve a value."""


@dataclass(frozen=True)
class ResolvedCredentials:
    """Fully resolved AEP credential set for a single profile.

    Field names mirror the shape used by AEP's IMS Server-to-Server OAuth
    and the standard AEP API request headers (x-api-key, x-gw-ims-org-id,
    x-sandbox-name, Authorization: Bearer <access_token>).

    NEVER serialize this object into a tool response, log line, or
    exception message. It should only ever be consumed by internal HTTP
    client code in this process.
    """

    profile_name: str
    base_url: str
    client_id: str
    client_secret: str
    api_key: str
    org_id: str
    tech_account_id: str
    sandbox: str
    scopes: tuple[str, ...]

    def __repr__(self) -> str:  # pragma: no cover - defensive redaction
        return f"ResolvedCredentials(profile_name={self.profile_name!r}, sandbox={self.sandbox!r}, <redacted>)"


def _profiles_manifest_path() -> Path:
    # Repo-root-relative by convention; overridable for tests/containers.
    override = os.environ.get("MCP_PROFILES_PATH")
    if override:
        return Path(override)
    return Path(__file__).resolve().parents[6] / "secrets" / "profiles.json"


def _load_profiles_manifest() -> dict[str, Any]:
    path = _profiles_manifest_path()
    if not path.exists():
        raise SecretBackendError(
            f"Profiles manifest not found at {path}. Copy "
            "secrets/profiles.example.json to secrets/profiles.json and fill "
            "it in for local development."
        )
    with path.open("r", encoding="utf-8") as fh:
        return json.load(fh)


def _env_var_for(profile_name: str, field: str) -> str:
    """Deterministic env var name for a given profile + field, local-dev only.

    e.g. profile "cit_bank_test_user", field "client_secret" ->
    AEP_PROFILE_CIT_BANK_TEST_USER_CLIENT_SECRET
    """
    normalized = profile_name.upper().replace("-", "_")
    return f"AEP_PROFILE_{normalized}_{field.upper()}"


def _resolve_local(profile_name: str, profile_meta: dict[str, Any]) -> ResolvedCredentials:
    required_fields = [
        "base_url",
        "client_id",
        "client_secret",
        "api_key",
        "org_id",
        "tech_account_id",
        "sandbox",
    ]
    values: dict[str, str] = {}
    missing: list[str] = []
    for field in required_fields:
        env_name = _env_var_for(profile_name, field)
        value = os.environ.get(env_name)
        if not value:
            missing.append(env_name)
            continue
        values[field] = value

    if missing:
        raise SecretBackendError(
            f"Missing local secret env vars for profile '{profile_name}': "
            f"{', '.join(missing)}. Set them or switch MCP_SECRETS_BACKEND=vault."
        )

    scopes_raw = os.environ.get(_env_var_for(profile_name, "scopes"), "")
    scopes = tuple(s.strip() for s in scopes_raw.split(",") if s.strip())

    return ResolvedCredentials(
        profile_name=profile_name,
        base_url=values["base_url"],
        client_id=values["client_id"],
        client_secret=values["client_secret"],
        api_key=values["api_key"],
        org_id=values["org_id"],
        tech_account_id=values["tech_account_id"],
        sandbox=values["sandbox"],
        scopes=scopes,
    )


def _resolve_vault(profile_name: str, profile_meta: dict[str, Any]) -> ResolvedCredentials:
    vault_ref = profile_meta.get("vaultRef")
    if not vault_ref:
        raise SecretBackendError(f"Profile '{profile_name}' has no vaultRef configured.")
    raise NotImplementedError(
        f"wire up Vault/AWS Secrets Manager lookup for vaultRef={vault_ref!r} here "
        "(e.g. hvac.Client().secrets.kv.v2.read_secret_version, or boto3 secretsmanager "
        "get_secret_value) and map the returned fields onto ResolvedCredentials"
    )


def resolve_profile(profile_name: str, domain: str) -> ResolvedCredentials:
    """Resolve a profile name + target domain into usable AEP credentials.

    Args:
        profile_name: Logical profile identifier, e.g. "cit_bank_test_user".
            Tool callers pass this string; they never pass secret material.
        domain: The domain/tenant context the credential will be used
            against (e.g. "citbank.com"). Validated against the profile's
            `allowedDomains` so a credential can't be replayed against an
            unintended target.

    Returns:
        A ResolvedCredentials instance for internal use only.

    Raises:
        ProfileNotFoundError: profile_name is not in the manifest.
        DomainNotAllowedError: domain is not in the profile's allowlist.
        SecretBackendError: the configured backend could not resolve values.
    """
    manifest = _load_profiles_manifest()
    profile_meta = manifest.get(profile_name)
    if profile_meta is None:
        raise ProfileNotFoundError(f"Unknown profile: {profile_name!r}")

    allowed_domains: list[str] = profile_meta.get("allowedDomains", [])
    if domain not in allowed_domains and "*" not in allowed_domains:
        raise DomainNotAllowedError(
            f"Profile {profile_name!r} is not authorized for domain {domain!r}. "
            f"Allowed: {allowed_domains}"
        )

    backend = os.environ.get("MCP_SECRETS_BACKEND", "local").lower()
    if backend == "local":
        return _resolve_local(profile_name, profile_meta)
    if backend == "vault":
        return _resolve_vault(profile_name, profile_meta)

    raise SecretBackendError(f"Unknown MCP_SECRETS_BACKEND: {backend!r}")
